README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# {coloratio}

<!-- badges: start -->

[![Project Status: Concept – Minimal or no implementation has been done
yet, or the repository is only intended to be a limited example, demo,
or
proof-of-concept.](https://www.repostatus.org/badges/latest/concept.svg)](https://www.repostatus.org/#concept)
[![R-CMD-check](https://github.com/matt-dray/coloratio/workflows/R-CMD-check/badge.svg)](https://github.com/matt-dray/coloratio/actions)
[![Codecov test
coverage](https://codecov.io/gh/matt-dray/coloratio/branch/main/graph/badge.svg)](https://codecov.io/gh/matt-dray/coloratio?branch=main)
[![CodeFactor](https://www.codefactor.io/repository/github/matt-dray/coloratio/badge/main)](https://www.codefactor.io/repository/github/matt-dray/coloratio/overview/main)
[![rostrum.blog
post](https://img.shields.io/badge/rostrum.blog-post-008900?style=flat&labelColor=black&logo=data:image/gif;base64,R0lGODlhEAAQAPEAAAAAABWCBAAAAAAAACH5BAlkAAIAIf8LTkVUU0NBUEUyLjADAQAAACwAAAAAEAAQAAAC55QkISIiEoQQQgghRBBCiCAIgiAIgiAIQiAIgSAIgiAIQiAIgRAEQiAQBAQCgUAQEAQEgYAgIAgIBAKBQBAQCAKBQEAgCAgEAoFAIAgEBAKBIBAQCAQCgUAgEAgCgUBAICAgICAgIBAgEBAgEBAgEBAgECAgICAgECAQIBAQIBAgECAgICAgICAgECAQECAQICAgICAgICAgEBAgEBAgEBAgICAgICAgECAQIBAQIBAgECAgICAgIBAgECAQECAQIBAgICAgIBAgIBAgEBAgECAgECAgICAgICAgECAgECAgQIAAAQIKAAAh+QQJZAACACwAAAAAEAAQAAAC55QkIiESIoQQQgghhAhCBCEIgiAIgiAIQiAIgSAIgiAIQiAIgRAEQiAQBAQCgUAQEAQEgYAgIAgIBAKBQBAQCAKBQEAgCAgEAoFAIAgEBAKBIBAQCAQCgUAgEAgCgUBAICAgICAgIBAgEBAgEBAgEBAgECAgICAgECAQIBAQIBAgECAgICAgICAgECAQECAQICAgICAgICAgEBAgEBAgEBAgICAgICAgECAQIBAQIBAgECAgICAgIBAgECAQECAQIBAgICAgIBAgIBAgEBAgECAgECAgICAgICAgECAgECAgQIAAAQIKAAA7)](https://www.rostrum.blog/2020/12/30/coloratio/)
<!-- badges: end -->

This work-in-progress R package helps calculate color-contrast values so
you can make decisions to improve the visual accessibility of your R
outputs. You can [read a blog post about
it](https://www.rostrum.blog/2020/12/30/coloratio/).

It was originally developed with the need for selecting high-contrast
text colors for use on block-colored backgrounds. The ultimate aim is to
create a tool for compliance with [WCAG
3.0](https://w3c.github.io/silver/guidelines/).

Feel free to add an issue or a make a pull request. Please follow [the
code of
conduct](https://www.contributor-covenant.org/version/2/0/code_of_conduct/).

## Similar tools

[The {savonliquide}
package](https://github.com/feddelegrand7/savonliquide) by [Ihaddaden M.
El Fodil](https://twitter.com/moh_fodil) achieves the main aim of this
package using [the WebAIM contrast checker
API](https://webaim.org/resources/contrastchecker/). It’s [available on
CRAN](https://CRAN.R-project.org/package=savonliquide).

## Installation

You can install the development version of {coloratio} from GitHub with:

``` r
remotes::install_github("matt-dray/coloratio")
```

## Examples

### Get contrast ratio

The `cr_get_ratio()` function returns the color-contrast ratio of two
user-provided colors. You can provide hex values, named colors, or both.

``` r
library(coloratio)
cr_get_ratio("#000000", "white")
# [1] 21
```

You should aim for a value of 4.5 or greater. You’ll get a warning if
the contrast between the colors is insufficient.

``` r
cr_get_ratio("black", "grey10")
# Warning in cr_get_ratio("black", "grey10"): Aim for a value of 4.5 or higher.
# [1] 1.206596
```

[Read more about the
calculation](https://www.w3.org/TR/WCAG/#dfn-relative-luminance) in the
WCAG 2.1 guidance.

### Color contrast demo

`cr_get_ratio()` also lets you view a plot of text on a block background
using the supplied colors by setting `view = TRUE`.

``` r
cr_get_ratio("#000000", "olivedrab", view = TRUE)
```

<img src="man/figures/README-demo-example-1.png" width="50%" />

    # [1] 5.51863

You can access this visual functionality in isolation with
`cr_view_contrast()`.

``` r
cr_view_contrast("blue4", "sienna1")
```

<img src="man/figures/README-demo-2-example-1.png" width="50%" />

### Choose black or white

You may want to alter programmatically the color of overlaying elements
when the underlying color varies. For example, when putting text over
the bars of a grouped bar chart.

The function `cr_choose_bw()` returns `"white"` or `"black"` depending
on which has the better contrast ratio with the user-supplied color.

``` r
# Supply a character vector, receive a character vector
cr_choose_bw(c("gray10", "gray90", "lightyellow"))
# [1] "white" "black" "black"
```

Black is chosen in the event of a tie.

### Choose a high-contrast color

You can use `cr_choose_color()` to check a user-supplied color against
all named R colors and return those with greatest contrast.

``` r
cr_choose_color("midnightblue")
# [1] "white"

cr_choose_color(
  col = "olivedrab",  # user-supplied color
  n = 3,              # number of colors to return
  ex_bw = TRUE        # exclude black, whites, grays?
)
# [1] "navy"     "navyblue" "blue4"
```

You have the option to exclude black, whites and grays from being
returned (via the `ex_bw` argument) because they’re boring.

## Code of Conduct

Please note that the {coloratio} project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# coloratio 0.0.0.9004

* Corrected regex for hex values (thanks to @EllaKaye).
* Vectorised `cr_choose_bw()` (#12) and gave README example of vector input.
* Change 'colour' to 'color' throughout.

# coloratio 0.0.0.9003

* Added `cr_view_contrast()` function (#2).
* Added tests.
* Added Codecov badge and GitHub Action for coverage.
* Added CodeFactor badge.

# coloratio 0.0.0.9002

* Added `cr_choose_color()` function.

# coloratio 0.0.0.9001

* Added `cr_choose_bw()` function (#1).
* Updated naming convention to `cr_*()` (#9).
* Updated 'color' throughout (#11).
* Added `cr_choose_bw()` to README.
* Added code of conduct.
* Add basic stopifnots.

# coloratio 0.0.0.9000

* Added `get_ratio()` function.
* Added NEWS, LICENSE, README.


tests/testthat.R
library(testthat)
library(coloratio)

test_check("coloratio")


tests/testthat/test-contrast.R
test_that("cr_get_ratio() expected outputs are returned", {

  expect_type(cr_get_ratio("black", "white"), "double")
  expect_type(cr_get_ratio("black", "white", view = TRUE), "double")

})

test_that("cr_choose_bw() expected outputs are returned", {

  expect_type(cr_choose_bw("black"), "character")
  expect_identical(cr_choose_bw("black"), "white")
  expect_identical(cr_choose_bw("white"), "black")

  expect_type(cr_choose_bw(c("black", "white")), "character")
  expect_identical(cr_choose_bw(c("black", "white")), c("white", "black"))
  expect_length(cr_choose_bw(c("black", "white")), 2)

  expect_type(cr_choose_bw(list("black", "white")), "character")
  expect_identical(cr_choose_bw(list("black", "white")), c("white", "black"))
  expect_length(cr_choose_bw(list("black", "white")), 2)

})

test_that("cr_choose_color() expected outputs are returned", {

  expect_type(cr_choose_color("black"), "character")
  expect_length(cr_choose_color("black", n = 3), 3)
  expect_setequal(
    cr_choose_color("black", n = 3), c("white", "gray100", "grey100")
  )
  expect_setequal(
    cr_choose_color("black", n = 3, ex_bw = TRUE),
    c("ivory", "ivory1", "lightyellow")
  )

})

test_that("cr_view_contrast() expected outputs are returned", {

  expect_silent(cr_view_contrast("black", "white"))  # produces a plot

})

test_that("cr_get_ratio() errors/warnings as expected", {

  expect_error(cr_get_ratio("blurple", "white"))  # not a real color
  expect_error(cr_get_ratio("000000", "white"))  # not preceded by #
  expect_error(cr_get_ratio("black", "white", "gray"))  # too many colors
  expect_error(cr_get_ratio(1, "white"))  # numeric not accepted
  expect_error(cr_get_ratio(TRUE, "white"))  # logical not accepted

  expect_warning(cr_get_ratio("black", "black"))  # low contrast
  expect_silent(cr_get_ratio("black", "black", quiet = TRUE))  # no warning

})

test_that("cr_choose_bw() errors as expected", {

  expect_error(cr_choose_bw("blurple"))  # not a real color
  expect_error(cr_choose_bw("000000"))  # not preceded by #
  expect_error(cr_choose_bw("black", "white"))  # too many colors
  expect_error(cr_choose_bw(1))  # numeric not accepted
  expect_error(cr_choose_bw(TRUE))  # logical not accepted

})

test_that("cr_choose_color() errors as expected", {

  expect_error(cr_choose_color("blurple"))  # not a real color
  expect_error(cr_choose_color("000000"))  # not preceded by #
  expect_error(cr_choose_color("black", "white"))  # too many colors
  expect_error(cr_choose_color(1))  # numeric not accepted
  expect_error(cr_choose_color(TRUE))  # logical not accepted

  expect_error(cr_choose_color("black", n = "x"))  # n must be numeric
  expect_error(cr_choose_color("black", n = -1))  # n must be >=1
  expect_error(cr_choose_color("black", ex_bw = "x"))  # ex_bw must be logical
  expect_error(cr_choose_color("black", ex_bw = 1))  # ex_bw must be logical

})

test_that("cr_view_contrast() errors as expected", {

  expect_error(cr_view_contrast("blurple", "white"))  # not a real color
  expect_error(cr_view_contrast("000000", "white"))  # not preceded by #
  expect_error(cr_view_contrast("black", "white", "gray"))  # too many colors
  expect_error(cr_view_contrast(1, "white"))  # numeric not accepted
  expect_error(cr_view_contrast(TRUE, "white"))  # logical not accepted

})


R/contrast.R
#' Get Contrast Ratio of Two Colors
#'
#' Calculate the color contrast ratio of two provided colors. Intended as an
#' visual accessibility aid when selecting a text color to place over a single-
#' color background. The output value should be 4.5 or higher to ensure
#' sufficient contrast and readability. This function is currently based on
#' \href{https://www.w3.org/TR/WCAG21/}{WCAG 2.1}. See
#' \href{w3.org/TR/WCAG/#dfn-relative-luminance}{WCAG} for more information
#' about the calculation.
#'
#' @param col_1 Six-digit hex value preceded by '#', or a named color from
#'     \code{\link[grDevices]{colors}}
#' @param col_2 Six-digit hex value preceded by '#', or a named color from
#'     \code{\link[grDevices]{colors}}
#' @param quiet Whether to print warning when the ratio value is lower than 4.5.
#' @param view Whether to plot a demonstration of col_1 text on a col_2
#'     background, and vice versa, for visual inspection. Uses
#'     \code{\link{cr_view_contrast}}.
#'
#' @return A double.
#' @export
#'
#' @examples cr_get_ratio("#FFFFFF", "white")

cr_get_ratio <- function(col_1, col_2, quiet = FALSE, view = FALSE) {

  if(
    (!grepl("^#", col_1) & !col_1 %in% grDevices::colors() |
     !grepl("^#", col_2) & !col_2 %in% grDevices::colors()) |
    (grepl("^#", col_1) & !grepl("^#[0-9a-fA-F]{6}$", col_1) |
     grepl("^#", col_2) & !grepl("^#[0-9a-fA-F]{6}$", col_2))
  ) {
    stop('Inputs must be in colors() if named, or of the hex form "#RRGGBB".\n')
  }

  if (class(quiet) != "logical") {
    stop("Argument 'quiet' must be TRUE or FALSE.\n")
  }

  # Convert colous to RGB and scale 0 to 1
  d <- t(grDevices::col2rgb(c(col_1, col_2))) / 255

  # Convert value
  d <- apply(
    d, 2, function(x) ifelse(
      x <= 0.03928, x / 12.92, ((x + 0.055) / 1.055) ^ 2.4
    )
  )

  # Calculate luminance values
  d <- as.data.frame(d)
  d$L <- (0.2126 * d$red) + (0.7152 * d$green) + (0.0722 * d$blue)

  # Calculate contrast ratio
  d <- d[order(d$L), ]
  cr <- (d[2, "L"] + 0.05) / (d[1, "L"] + 0.05)

  # Warn if contrast threshold not reached
  if (!quiet & cr <= 4.5) {
    warning("Aim for a value of 4.5 or higher.\n")
  }

  # Plot contrast samples
  if (view) {
    cr_view_contrast(col_1, col_2)
  }

  return(cr)

}

#' Choose White or Black to Overlay On a Supplied Background Color
#'
#' Selects whether black or white has the greater contrast with a user-
#' supplied color. Useful for choosing a text color to overlay on a block-color
#' background, like value labels over the bars of a bar chart. Calculated as per
#' \code{\link{cr_get_ratio}}. Defaults to black in the case of a tie.
#'
#' @param col_bg A character vector of colors against which to select either
#'     black or white, whichever has maximum contrast. Supply colors as
#'     six-digit hex values preceded by '#', or named colors from
#'     \code{\link[grDevices]{colors}}.
#'
#' @return A character vector of values "black" or "white". The length matches
#'     the input.
#'
#' @export
#'
#' @examples cr_choose_bw(c("white", "gray90", "gray50", "gray10", "black"))

cr_choose_bw <- function(col_bg) {

  # Choose "black" or "white" for each element of col_bg, return vector
  vapply(
    col_bg,
    .choose_bw_one,
    character(1),
    USE.NAMES = FALSE
  )

}

#' Choose a High-Contrast Color for a Given Color
#'
#' Given a user-supplied color, what's a good color to pair it with for maximum
#' contrast? Compares provided color against all named R colors, as per
#' \code{\link[grDevices]{colors}}. Contrast calculated as per
#' \code{\link{cr_get_ratio}}.
#'
#' @param col A character value representing a color. Can be a six-digit hex
#'     value preceded by '#', or a named color from
#' @param n Number of named colors to return. Color with highest contrast
#'     is returned first.
#' @param ex_bw Exclude black and variants of white and gray variants?
#'
#' @return A character value that's a named R color.
#' @export
#'
#' @examples cr_choose_color("lightyellow")

cr_choose_color <- function(col, n = 1, ex_bw = FALSE) {

  if (class(n) != "numeric" | n < 1) {
    stop("'n' must be a positive numeric value.\n")
  }

  if (class(ex_bw) != "logical") {
    stop("Argument 'ex_bw' must be TRUE or FALSE.\n")
  }

  # Calculate all contrast ratios against the named colors
  ratios <- sapply(
    grDevices::colors(),
    function(x) cr_get_ratio(x, col, quiet = TRUE)
  )

  # Get the name of the result with the highest ratio
  if (!ex_bw) {
    result <- names(ratios[order(-ratios)][1:n])
  } else if (ex_bw) {
    bw_regex <-  "black|white|grey|gray"
    ratios_ex_bw <- ratios[!grepl(bw_regex, names(ratios))]
    result <- names(ratios_ex_bw[order(-ratios_ex_bw)][1:n])
  }

  return(result)

}

#' Plot a Demo of User-Supplied Color Pair
#'
#' Plots text of one color on a background of another color, and vice versa.
#' Used to visualise contrasts.
#'
#' @param col_1 Six-digit hex value preceded by '#', or a named color from
#'     \code{\link[grDevices]{colors}}.
#' @param col_2 Six-digit hex value preceded by '#', or a named color from
#'     \code{\link[grDevices]{colors}}.
#'
#' @return Character value or vector.
#' @export
#'
#' @examples cr_view_contrast("yellow", "black")

cr_view_contrast <- function(col_1, col_2) {

  if(
    (!grepl("^#", col_1) & !col_1 %in% grDevices::colors() |
     !grepl("^#", col_2) & !col_2 %in% grDevices::colors()) |
    (grepl("^#", col_1) & !grepl("^#[0-9a-fA-F]{6}$", col_1) |
     grepl("^#", col_2) & !grepl("^#[0-9a-fA-F]{6}$", col_2))
  ) {
    stop('Inputs must be in colors() if named, or of the hex form "#RRGGBB".\n')
  }

  # Reduce plot margins
  graphics::par(mar = rep(1, 4))

  # Plot stacked bar
  graphics::barplot(
    matrix(c(1, 1)),
    col = c(col_2, col_1),
    border = "white",
    yaxt = "n"
  )

  # Add overlaying text
  graphics::text(0.7, 0.5, paste(col_1, "\non", col_2), col = col_1, cex = 3)
  graphics::text(0.7, 1.5, paste(col_2, "\non", col_1), col = col_2, cex = 3)

}
